sinabs/backend/dynapcnn/sinabs_edges_handler.py

```diff
@@ -0,0 +1,472 @@
+"""
+functionality : functions implementing the pre-processing of edges into blocks of nodes (modules) for future
+                creation of DynapcnnLayer objects.
+author        : Willian Soares Girao
+contact       : williansoaresgirao@gmail.com
+"""
+
+from collections import deque
+from typing import Dict, List, Set, Tuple, Type
+
+from torch import Size, nn
+
+from .connectivity_specs import VALID_SINABS_EDGE_TYPES
+from .exceptions import (
+    InvalidEdge,
+    InvalidGraphStructure,
+    UnmatchedNode,
+    UnmatchedPoolingEdges,
+)
+from .utils import Edge
+
+
+def collect_dynapcnn_layer_info(
+    indx_2_module_map: Dict[int, nn.Module],
+    edges: Set[Edge],
+    nodes_io_shapes: Dict[int, Dict[str, Tuple[Size, Size]]],
+    entry_nodes: Set[int],
+) -> Dict[int, Dict]:
+    """Collect information to construct DynapcnnLayer instances.
+
+    Validate and sort edges based on the type of nodes they connect.
+    Iterate over edges in order of their type. For each neuron->weight edge
+    generate a new dict to collect information for the corresponding dynapcnn layer.
+    Then add pooling based on neuron->pooling type edges. Collect additional pooling
+    from pooling->pooling type edges. Finally set layer destinations based on
+    neuron/pooling->weight type of edges.
+
+    Parameters
+    ----------
+    indx_2_module_map (dict): Maps node IDs of the graph as `key` to their associated module as `value`
+    edges (set of tuples): Represent connections between two nodes in computational graph
+    nodes_io_shapes (dict): Map from node ID to dict containing node's in- and output shapes
+    entry_nodes (set of int): IDs of nodes that receive external input
+
+    Returns
+    -------
+    dynapcnn_layer_info (dict): Each 'key' is the index of a future 'DynapcnnLayer' and
+        'value' is a dictionary, with keys 'conv', 'neuron', and 'destinations',
+        containing corresponding node ids and modules required to build the layer
+    """
+    # TODO: Handle DVS layer
+
+    # Sort edges by edge type (type of layers they connect)
+    edges_by_type: Dict[str, Set[Edge]] = sort_edges_by_type(
+        edges=edges, indx_2_module_map=indx_2_module_map
+    )
+
+    if "weight-neuron" not in edges_by_type:
+        raise InvalidGraphStructure(
+            "Any dynapcnn layer must contain a weight layer (e.g. Conv2d, Linear) "
+            "that is directly connected to a neuron layer (e.g. IAFSqueeze). "
+            "None such weight-neuron pair has been found in the provided network."
+        )
+
+    # Dict to collect information for each future dynapcnn layer
+    dynapcnn_layer_info = dict()
```

Comment:
The function returns this variable but the variable name receiving the return of this function is `dcnnl_map`: should it also perhaps be renamed to `dynapcnn_layer_info` to maintain nomeclature consistency?

Reply:
Yes good point. Feel free to chose a name. Maybe `dcnnl_info` because it's shorter?  We should then make it consistent across different files